File: 0014_longestCommonPrefix/python/Solution.py
class Solution:
    def longestCommonPrefix(self, strs: list[str]) -> str:
        if len(strs) == 0:
            return ""
        prefix = strs[0]
        for word in strs[1:]:
            while word.find(prefix) != 0:
                prefix = prefix[0:-1]
                if prefix == "":
                    return ""
        return prefix


if __name__ == "__main__":
    s = Solution()
    print(s.longestCommonPrefix(["flower", "flow", "flight"]))
    print(s.longestCommonPrefix(["dog", "racecar", "car"]))
    print(s.longestCommonPrefix(["c", "c"]))
    print(s.longestCommonPrefix(["a"]))
    print(s.longestCommonPrefix([]))
    print(s.longestCommonPrefix(["c","acc","ccc"]))

File: 0014_longestCommonPrefix/cpp/Solution.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        if (strs.empty()) {
            return "";
        }
        string prefix = strs[0];
        for (int i = 1; i < strs.size(); ++i) {
            // std::cout << "strs[" << i << "] = " << strs[i] << std::endl;
            while (strs[i].find(prefix) != 0) {
                prefix.pop_back();
                // std::cout << "prefix: " << prefix << std::endl;
                if (prefix.empty()) {
                    return "";
                }
            }
        }
        return prefix;

    }
};

int main(int argc, char const *argv[])
{
    /* code */
    Solution s;
    vector<string> strs = {"flower","flow","flight"};
    cout << s.longestCommonPrefix(strs) << endl;

    strs = {"dog","racecar","car"};
    cout << s.longestCommonPrefix(strs) << endl;

    return 0;
}
